File: rpi_system/vision/classifier.py
"""
TFLite CNN classifier for animal detection on Raspberry Pi.
Optimized for tflite-runtime or tensorflow-lite on ARM architectures.
Includes mock fallback when running in testing environments without trained model weights.
"""
import os
import logging
from typing import List, Tuple, Optional
import numpy as np

logger = logging.getLogger("EdgeAI.Classifier")

class AnimalClassifier:
    """Handles TFLite model loading, image preprocessing, and inference."""

    def __init__(
        self,
        model_path: str,
        labels_path: str,
        confidence_threshold: float = 0.50,
        enable_roi_crop: bool = True,
        roi_crop_scale: float = 1.0,
        num_threads: int = 4,
    ):
        self.model_path = model_path
        self.labels_path = labels_path
        self.confidence_threshold = confidence_threshold
        self.enable_roi_crop = enable_roi_crop
        self.roi_crop_scale = roi_crop_scale
        self.num_threads = num_threads

        self.interpreter = None
        self.input_details = None
        self.output_details = None
        self.labels: List[str] = []
        self.is_mock_mode = False

        self._load_labels()
        self._load_model()

    def _load_labels(self):
        """Loads class labels from labels.txt."""
        if os.path.exists(self.labels_path):
            try:
                with open(self.labels_path, "r", encoding="utf-8") as f:
                    self.labels = [line.strip() for line in f if line.strip()]
                logger.info(f"Loaded {len(self.labels)} labels: {self.labels}")
            except Exception as e:
                logger.error(f"Error reading labels file: {e}")
                self.labels = ["Cat", "Dog", "Bird", "Wild Animal"]
        else:
            logger.warning(f"Labels file not found at {self.labels_path}. Using default classes.")
            self.labels = ["Cat", "Dog", "Bird", "Wild Animal"]

    def _load_model(self):
        """Loads TFLite interpreter from tflite-runtime or tensorflow."""
        # Check if file exists and has actual model bytes
        if not os.path.exists(self.model_path) or os.path.getsize(self.model_path) < 1024:
            logger.warning(
                f"Model file at '{self.model_path}' is missing or is an empty placeholder (< 1KB). "
                "Switching to SIMULATION MODE for testing triggers and display."
            )
            self.is_mock_mode = True
            return

        # Attempt to import tflite_runtime or ai_edge_litert first
        interpreter_cls = None
        try:
            from ai_edge_litert.interpreter import Interpreter
            interpreter_cls = Interpreter
            logger.info("Using ai_edge_litert.interpreter.")
        except ImportError:
            try:
                from tflite_runtime.interpreter import Interpreter
                interpreter_cls = Interpreter
                logger.info("Using tflite_runtime.interpreter.")
            except ImportError:
                try:
                    import tensorflow as tf
                    interpreter_cls = tf.lite.Interpreter
                    logger.info("Using tensorflow.lite.Interpreter.")
                except ImportError:
                    logger.warning(
                        "Neither tflite-runtime, ai-edge-litert, nor tensorflow is installed. "
                        "Inference will run in SIMULATION MODE."
                    )
                    self.is_mock_mode = True
                    return

        try:
            self.interpreter = interpreter_cls(model_path=self.model_path, num_threads=self.num_threads)
            self.interpreter.allocate_tensors()
            self.input_details = self.interpreter.get_input_details()
            self.output_details = self.interpreter.get_output_details()
            logger.info(
                f"Successfully loaded TFLite model. "
                f"Input shape: {self.input_details[0]['shape']}, "
                f"Input type: {self.input_details[0]['dtype']}"
            )
        except Exception as e:
            logger.error(f"Failed to load TFLite model from '{self.model_path}': {e}")
            logger.warning("Falling back to SIMULATION MODE.")
            self.is_mock_mode = True

    @staticmethod
    def crop_roi(
        frame: np.ndarray,
        roi_scale: float = 1.0,
    ) -> Tuple[np.ndarray, Tuple[int, int, int, int]]:
        """
        Crops a square Region of Interest (ROI) from the center of the frame.
        Prevents 16:9 to 1:1 squishing distortion and focuses on the active detection zone.

        Returns:
            (cropped_frame, (x1, y1, x2, y2))
        """
        h, w = frame.shape[:2]
        min_dim = min(h, w)
        crop_size = int(min_dim * max(0.1, min(1.0, roi_scale)))

        x1 = max(0, (w - crop_size) // 2)
        y1 = max(0, (h - crop_size) // 2)
        x2 = x1 + crop_size
        y2 = y1 + crop_size

        return frame[y1:y2, x1:x2], (x1, y1, x2, y2)

    def preprocess_image(self, frame: np.ndarray) -> np.ndarray:
        """
        Extracts square ROI, resizes, and normalizes the input frame for the model.
        Supports both UINT8 and FLOAT32 input tensors.
        """
        import cv2

        # Apply ROI crop to preserve aspect ratio
        if self.enable_roi_crop:
            cropped_frame, _ = self.crop_roi(frame, self.roi_crop_scale)
        else:
            cropped_frame = frame

        if self.input_details:
            _, height, width, channels = self.input_details[0]["shape"]
            input_dtype = self.input_details[0]["dtype"]
        else:
            height, width, channels = 224, 224, 3
            input_dtype = np.float32

        # Convert BGR to RGB and resize to model input shape
        rgb_frame = cv2.cvtColor(cropped_frame, cv2.COLOR_BGR2RGB)
        resized_frame = cv2.resize(rgb_frame, (width, height))

        # Add batch dimension
        input_data = np.expand_dims(resized_frame, axis=0)

        if input_dtype == np.float32:
            input_data = input_data.astype(np.float32)
        elif input_dtype == np.uint8:
            input_data = input_data.astype(np.uint8)

        return input_data

    def predict(self, frame: np.ndarray) -> Tuple[str, float]:
        """
        Performs inference on a camera frame.
        Returns: (label_name, confidence_score)
        """
        if self.is_mock_mode or self.interpreter is None:
            # Demo inference when model weights are pending
            import random
            mock_class = random.choice(self.labels) if self.labels else "Animal"
            mock_conf = round(random.uniform(0.75, 0.98), 2)
            logger.info(f"[SIMULATION] Predicted: {mock_class} ({mock_conf * 100:.1f}%)")
            return mock_class, mock_conf

        try:
            input_data = self.preprocess_image(frame)
            self.interpreter.set_tensor(self.input_details[0]["index"], input_data)
            self.interpreter.invoke()

            output_data = self.interpreter.get_tensor(self.output_details[0]["index"])
            probabilities = np.squeeze(output_data)

            # Dequantize if model output is quantized
            if self.output_details[0]["dtype"] == np.uint8:
                scale, zero_point = self.output_details[0]["quantization"]
                if scale > 0:
                    probabilities = scale * (probabilities.astype(np.float32) - zero_point)

            top_idx = int(np.argmax(probabilities))
            confidence = float(probabilities[top_idx])

            label = (
                self.labels[top_idx]
                if top_idx < len(self.labels)
                else f"Class_{top_idx}"
            )

            logger.info(f"Prediction: {label} (confidence: {confidence:.2%})")
            return label, confidence

        except Exception as e:
            logger.error(f"Inference error: {e}")
            return "Unknown", 0.0

File: rpi_system_C++/src/vision/classifier.cpp
#include "../../include/vision/classifier.hpp"
#include "../../include/config.hpp"
#include <filesystem>
#include <fstream>
#include <opencv2/opencv.hpp>
#include <optional>
#include <spdlog/fmt/ranges.h>
#include <spdlog/spdlog.h>
#include <string>
#include <tensorflow/lite/interpreter.h>
#include <tensorflow/lite/kernels/register.h>
#include <tensorflow/lite/model.h>
#include <tuple>
#include <vector>

AnimalClassifier::AnimalClassifier()
    : model_path(Config::MODEL_PATH), labels_path(Config::LABELS_PATH),
      confidence_threshold(Config::CONFIDENCE_THRESHOLD),
      enable_roi_crop(Config::ENABLE_ROI_CROP), roi_crop_scale(Config::ROI_CROP_SCALE) {
  load_labels();
  load_model();
}

AnimalClassifier::~AnimalClassifier() {}

void AnimalClassifier::load_labels() {
  if (std::filesystem::exists(labels_path)) {
    try {
      std::ifstream file(labels_path);
      std::string line;
      while (std::getline(file, line)) {
        if (!line.empty()) {
          labels.push_back(line);
        }
      }
      spdlog::info("Loaded {} labels: {}", labels.size(), labels);
    } catch (const std::exception &e) {
      spdlog::error("Error reading labels file: {}", e.what());
      labels = {"Cat", "Dog", "Bird", "Wild Animal"};
    }
  } else {
    spdlog::warn("Labels file not found at {}. Using default classes.",
                 labels_path);
    labels = { "Cat", "Dog", "Bird", "Wild Animal" };
  }
}

void AnimalClassifier::load_model() {
  if (!std::filesystem::exists(model_path) ||
      !std::filesystem::is_regular_file(model_path)) {
    spdlog::warn(
        "Model file at '{}' is missing or is an empty placeholder (< 1KB). "
        "Switching to SIMULATION MODE for testing triggers and display.",
        model_path);
    is_mock_mode = true;
    return;
  }

  model = tflite::FlatBufferModel::BuildFromFile(model_path.c_str());
  if (!model) {
    spdlog::warn("Failed to load TFLite model from '{}'. Falling back to "
                 "SIMULATION MODE.",
                 model_path);
    is_mock_mode = true;
    return;
  }

  tflite::ops::builtin::BuiltinOpResolver resolver;
  tflite::InterpreterBuilder builder(*model, resolver);
  builder(&interpreter);

  if (interpreter) {
    interpreter->SetNumThreads(4); // Utilize all 4 cores on Raspberry Pi
  }

  if (!interpreter) {
    spdlog::error("Failed to construct TFLite interpreter. Falling back to "
                  "SIMULATION MODE.");
    is_mock_mode = true;
    return;
  }

  if (interpreter->AllocateTensors() != kTfLiteOk) {
    spdlog::error(
        "Failed to allocate tensors. Falling back to SIMULATION MODE.");
    is_mock_mode = true;
    return;
  }
  spdlog::info("Successfully loaded TFLite model.");
}

std::pair<cv::Mat, cv::Rect> AnimalClassifier::crop_roi(const cv::Mat &frame,
                                                        float roi_scale) {
  int h = frame.rows;
  int w = frame.cols;

  int min_dim = std::min(h, w);
  int crop_size = int(min_dim * std::max(0.1f, std::min(1.0f, roi_scale)));

  int x1 = std::max(0, (w - crop_size) / 2);
  int y1 = std::max(0, (h - crop_size) / 2);

  cv::Rect roi(x1, y1, crop_size, crop_size);
  cv::Mat cropped = frame(roi);

  return {cropped, roi};
}

cv::Mat AnimalClassifier::preprocess_image(const cv::Mat &frame) {
  cv::Mat cropped_frame;
  if (enable_roi_crop) {
    cropped_frame = crop_roi(frame, roi_crop_scale).first;
  } else {
    cropped_frame = frame;
  }

  TfLiteTensor *input = interpreter->input_tensor(0);
  int height = input->dims->data[1];
  int width = input->dims->data[2];
  TfLiteType input_dtype = input->type;

  cv::Mat rgb_frame;
  cv::cvtColor(cropped_frame, rgb_frame, cv::COLOR_BGR2RGB);

  cv::Mat resized_frame;
  cv::resize(rgb_frame, resized_frame, cv::Size(width, height));

  cv::Mat input_data;
  if (input_dtype == kTfLiteFloat32) {
    resized_frame.convertTo(input_data, CV_32FC3);
  } else {
    resized_frame.convertTo(input_data, CV_8UC3);
  }

  return input_data;
}

std::tuple<std::string, double>
AnimalClassifier::predict(const cv::Mat &frame) {
  if (is_mock_mode || !interpreter) {
    std::string mock_class =
        labels.empty() ? "Animal" : labels[rand() % labels.size()];
    double mock_conf = 0.75 + (rand() / (RAND_MAX / (0.98 - 0.75)));
    spdlog::info("[SIMULATION] Predicted: {} ({:.1f}%)", mock_class,
                 mock_conf * 100);
    return {mock_class, mock_conf};
  }

  try {
    cv::Mat input_data = preprocess_image(frame);
    TfLiteTensor *input_tensor = interpreter->input_tensor(0);

    size_t bytes = input_data.total() * input_data.elemSize();
    if (!input_data.isContinuous() || bytes != input_tensor->bytes) {
      spdlog::error("Input size mismatch: {} vs {}", bytes,
                    input_tensor->bytes);
      return {"Unknown", 0.0};
    }

    memcpy(input_tensor->data.raw, input_data.data, bytes);

    if (interpreter->Invoke() != kTfLiteOk) {
      spdlog::error("Inference error: Invoke failed.");
      return {"Unknown", 0.0};
    }

    TfLiteTensor *output_tensor = interpreter->output_tensor(0);
    if (output_tensor->dims->size < 2) {
      spdlog::error("Unexpected output shape (rank {})",
                    output_tensor->dims->size);
      return {"Unknown", 0.0};
    }
    int num_classes = output_tensor->dims->data[1];

    double max_conf = -1.0;
    int top_idx = 0;

    if (output_tensor->type == kTfLiteFloat32) {
      const float *output_data = output_tensor->data.f;
      for (int i = 0; i < num_classes; i++) {
        if (output_data[i] > max_conf) {
          max_conf = output_data[i];
          top_idx = i;
        }
      }
    } else if (output_tensor->type == kTfLiteUInt8) {
      const uint8_t *output_data = output_tensor->data.uint8;
      float scale = output_tensor->params.scale;
      int zero_point = output_tensor->params.zero_point;

      for (int i = 0; i < num_classes; i++) {
        double conf = scale * (output_data[i] - zero_point);
        if (conf > max_conf) {
          max_conf = conf;
          top_idx = i;
        }
      }
    } else {
      spdlog::error("Unsupported output tensor type: {}",
                    TfLiteTypeGetName(output_tensor->type));
      return {"Unknown", 0.0};
    }

    std::string label = (static_cast<size_t>(top_idx) < labels.size())
                            ? labels[top_idx]
                            : "Class_" + std::to_string(top_idx);
    spdlog::info("Prediction: {} (confidence: {:.2f}%)", label, max_conf * 100);
    return {label, max_conf};

  } catch (const std::exception &e) {
    spdlog::error("Inference error: {}", e.what());
    return {"Unknown", 0.0};
  }
}